solls/LanguageServer.cpp

```diff
@@ -0,0 +1,668 @@
+	result.capabilities.textDocumentSync.change = lsp::protocol::TextDocumentSyncKind::Incremental;
+	result.capabilities.definitionProvider = true; // go-to-definition feature
+
+	reply(_args.requestId, result);
+}
+
+void LanguageServer::operator()(lsp::protocol::InitializedNotification const&)
+{
+	// NB: this means the client has finished initializing. Now we could maybe start sending
+	// events to the client.
+	logMessage("LanguageServer: Client initialized");
+}
+
+void LanguageServer::operator()(lsp::protocol::DidOpenTextDocumentParams const& _args)
+{
+	logMessage("LanguageServer: Opening document: " + _args.textDocument.uri);
+
+	lsp::vfs::File const& file = m_vfs.insert(
+		_args.textDocument.uri,
+		_args.textDocument.languageId,
+		_args.textDocument.version,
+		_args.textDocument.text
+	);
+
+	validate(file);
+}
+
+void LanguageServer::operator()(lsp::protocol::DidChangeTextDocumentParams const& _didChange)
+{
+	if (lsp::vfs::File* file = m_vfs.find(_didChange.textDocument.uri); file != nullptr)
+	{
+		if (_didChange.textDocument.version.has_value())
+			file->setVersion(_didChange.textDocument.version.value());
+
+		for (lsp::protocol::TextDocumentContentChangeEvent const& contentChange: _didChange.contentChanges)
+		{
+			visit(util::GenericVisitor{
+				[&](lsp::protocol::TextDocumentRangedContentChangeEvent const& change) {
+#if !defined(NDEBUG)
+					ostringstream str;
+					str << "did change: " << change.range << " for '" << change.text << "'";
+					logMessage(str.str());
+#endif
+					file->modify(change.range, change.text);
+				},
+				[&](lsp::protocol::TextDocumentFullContentChangeEvent const& change) {
+					file->replace(change.text);
+				}
+			}, contentChange);
+		}
+
+		validate(*file);
+	}
+	else
+		logError("LanguageServer: File to be modified not opened \"" + _didChange.textDocument.uri + "\"");
+}
+
+void LanguageServer::operator()(lsp::protocol::DidCloseTextDocumentParams const& _didClose)
+{
+	logMessage("LanguageServer: didClose: " + _didClose.textDocument.uri);
+}
+
+void LanguageServer::validateAll()
+{
+	for (reference_wrapper<lsp::vfs::File const> const& file: m_vfs.files())
+		validate(file.get());
+}
+
+void LanguageServer::validate(lsp::vfs::File const& _file)
+{
+	PublishDiagnosticsList result;
+	validate(_file, result);
+
+	for (lsp::protocol::PublishDiagnosticsParams const& diag: result)
+		notify(diag);
+}
+
+frontend::ReadCallback::Result LanguageServer::readFile(string const& _kind, string const& _path)
+{
+	return m_fileReader->readFile(_kind, _path);
+}
+
+constexpr lsp::protocol::DiagnosticSeverity toDiagnosticSeverity(Error::Type _errorType)
+{
+	switch (_errorType)
+	{
+		case Error::Type::CodeGenerationError:
+		case Error::Type::DeclarationError:
+		case Error::Type::DocstringParsingError:
+		case Error::Type::ParserError:
+		case Error::Type::SyntaxError:
+		case Error::Type::TypeError:
+			return lsp::protocol::DiagnosticSeverity::Error;
```

Comment:
Hmm, we already make this severity distinction in the StandardCompiler. Perhaps we should add a severity enum to `Error::Type` instead.

Reply:
Even if we don't I think this function should be as simple as:
```
return (_errorType == Error::Type::Warning) ? lsp::protocol::DiagnosticSeverity::Warning : lsp::protocol::DiagnosticSeverity::Error;
```